File: gyaanquest/core/player.py
"""Player profile: XP, levels, titles, streaks and persistent stats."""
from __future__ import annotations

import json
from dataclasses import dataclass, field, asdict
from pathlib import Path

# (min_level, title) — checked top-down, first match wins.
LEVEL_TITLES = [
    (13, "Grandmaster"),
    (11, "Illuminated"),
    (9, "Mentor"),
    (7, "Pundit"),
    (5, "Scholar"),
    (3, "Apprentice"),
    (1, "Novice"),
]

# Total XP required to *be* a given level: 0, 100, 300, 600, 1000, ...
def xp_to_reach(level: int) -> int:
    return 50 * (level - 1) * level


def level_for_xp(total_xp: int) -> int:
    level = 1
    while xp_to_reach(level + 1) <= total_xp:
        level += 1
    return level


def title_for_level(level: int) -> str:
    for min_level, title in LEVEL_TITLES:
        if level >= min_level:
            return title
    return LEVEL_TITLES[-1][1]


@dataclass
class PlayerProfile:
    total_xp: int = 0
    total_answered: int = 0
    total_correct: int = 0
    best_streak: int = 0
    games_played: int = 0
    perfect_games: int = 0
    category_answered: dict = field(default_factory=dict)
    category_correct: dict = field(default_factory=dict)
    achievements: list = field(default_factory=list)

    # ---------------- levels ----------------
    @property
    def level(self) -> int:
        return level_for_xp(self.total_xp)

    @property
    def title(self) -> str:
        return title_for_level(self.level)

    def xp_progress(self):
        """Return (xp_into_current_level, xp_needed_for_next_level)."""
        current_floor = xp_to_reach(self.level)
        next_floor = xp_to_reach(self.level + 1)
        return self.total_xp - current_floor, next_floor - current_floor

    def add_xp(self, amount: int):
        """Add XP. Return (old_level, new_level)."""
        old = self.level
        self.total_xp += max(0, int(amount))
        return old, self.level

    # ---------------- stats ----------------
    def record_answer(self, category: str, correct: bool):
        self.total_answered += 1
        self.category_answered[category] = self.category_answered.get(category, 0) + 1
        if correct:
            self.total_correct += 1
            self.category_correct[category] = self.category_correct.get(category, 0) + 1

    def record_session(self, best_streak: int, perfect: bool):
        self.games_played += 1
        self.best_streak = max(self.best_streak, best_streak)
        if perfect:
            self.perfect_games += 1

    # ---------------- persistence ----------------
    def save(self, path) -> None:
        path = Path(path)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(json.dumps(asdict(self), indent=2))

    @classmethod
    def load(cls, path) -> "PlayerProfile":
        path = Path(path)
        if not path.exists():
            return cls()
        try:
            data = json.loads(path.read_text())
            valid = {f for f in cls.__dataclass_fields__}
            return cls(**{k: v for k, v in data.items() if k in valid})
        except (json.JSONDecodeError, TypeError, ValueError):
            return cls()

File: mobile/src/core/player.ts
import { PlayerProfileData } from '../types';
import { levelForXp, titleForLevel } from '../data/achievements';

export const DEFAULT_PROFILE: PlayerProfileData = {
  totalXp: 0,
  coins: 10000,
  totalAnswered: 0,
  totalCorrect: 0,
  bestStreak: 0,
  gamesPlayed: 0,
  perfectGames: 0,
  categoryAnswered: {},
  categoryCorrect: {},
  achievements: [],
};

export class PlayerProfile {
  data: PlayerProfileData;

  constructor(data: Partial<PlayerProfileData> = {}) {
    this.data = { ...DEFAULT_PROFILE, ...data };
  }

  get totalXp(): number {
    return this.data.totalXp;
  }

  set totalXp(value: number) {
    this.data.totalXp = value;
  }

  get coins(): number {
    return this.data.coins;
  }

  addCoins(amount: number): void {
    this.data.coins = Math.max(0, this.data.coins + amount);
  }

  get level(): number {
    return levelForXp(this.data.totalXp);
  }

  get title(): string {
    return titleForLevel(this.level);
  }

  get totalAnswered(): number {
    return this.data.totalAnswered;
  }

  get totalCorrect(): number {
    return this.data.totalCorrect;
  }

  get bestStreak(): number {
    return this.data.bestStreak;
  }

  set bestStreak(value: number) {
    this.data.bestStreak = Math.max(this.data.bestStreak, value);
  }

  get gamesPlayed(): number {
    return this.data.gamesPlayed;
  }

  get perfectGames(): number {
    return this.data.perfectGames;
  }

  get achievements(): string[] {
    return this.data.achievements;
  }

  get categoryCorrect(): Record<string, number> {
    return this.data.categoryCorrect;
  }

  xpProgress(): { into: number; needed: number } {
    const currentFloor = 50 * (this.level - 1) * this.level;
    const nextFloor = 50 * this.level * (this.level + 1);
    return {
      into: this.data.totalXp - currentFloor,
      needed: nextFloor - currentFloor,
    };
  }

  addXp(amount: number): { oldLevel: number; newLevel: number } {
    const old = this.level;
    this.data.totalXp += Math.max(0, Math.round(amount));
    return { oldLevel: old, newLevel: this.level };
  }

  recordAnswer(category: string, correct: boolean): void {
    this.data.totalAnswered += 1;
    this.data.categoryAnswered[category] =
      (this.data.categoryAnswered[category] ?? 0) + 1;
    if (correct) {
      this.data.totalCorrect += 1;
      this.data.categoryCorrect[category] =
        (this.data.categoryCorrect[category] ?? 0) + 1;
    }
  }

  recordSession(bestStreak: number, perfect: boolean): void {
    this.data.gamesPlayed += 1;
    this.data.bestStreak = Math.max(this.data.bestStreak, bestStreak);
    if (perfect) {
      this.data.perfectGames += 1;
    }
  }

  addAchievement(id: string): void {
    if (!this.data.achievements.includes(id)) {
      this.data.achievements.push(id);
    }
  }

  reset(): void {
    this.data = { ...DEFAULT_PROFILE };
  }

  toJSON(): PlayerProfileData {
    return { ...this.data };
  }

  static fromJSON(data: unknown): PlayerProfile {
    if (typeof data !== 'object' || data === null) {
      return new PlayerProfile();
    }
    const d = data as Partial<PlayerProfileData>;
    return new PlayerProfile({
      totalXp: d.totalXp ?? 0,
      coins: d.coins ?? 0,
      totalAnswered: d.totalAnswered ?? 0,
      totalCorrect: d.totalCorrect ?? 0,
      bestStreak: d.bestStreak ?? 0,
      gamesPlayed: d.gamesPlayed ?? 0,
      perfectGames: d.perfectGames ?? 0,
      categoryAnswered: d.categoryAnswered ?? {},
      categoryCorrect: d.categoryCorrect ?? {},
      achievements: d.achievements ?? [],
    });
  }
}

export { levelForXp, titleForLevel };
